src/librustc_typeck/check/coercion.rs

```diff
@@ -657,9 +664,8 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
                     );
                     match selcx.select(&obj_safe.with(pred.clone())) {
                         Ok(_) => (),
-                        Err(err) => {
-                            self.report_selection_error(&obj_safe, &err, false);
-                            return Err(TypeError::Mismatch)
+                        Err(_) => {
+                            return Err(TypeError::ObjectUnsafeCoercion(target_def_id))
```

Comment:
Ah, I guess this is the code I was recommending we delete (I still think we should delete it) -- but we could achieve something similar to this commit I suspect by adjusting the "cause" we wind up with.